src/XMakeBuildEngine/Definition/Project.cs

```diff
@@ -1097,28 +1106,14 @@ private List<GlobResult> GetAllGlobs(List<ProjectItemElement> projectItemElement
 
         private IEnumerable<GlobResult> GetAllGlobs(ProjectItemElement itemElement)
         {
-            Func<string, IElementLocation, HashSet<string>> expandItemSpecIntoFragments = (s, l) =>
-                {
-                    var expandedItemSpec = _data.Expander.ExpandIntoStringListLeaveEscaped(s, ExpanderOptions.ExpandProperties, l);
-
-                    // don't care about duplicates
-                    var set = new HashSet<string>(expandedItemSpec);
+            var includeItemspec = new ItemSpec<ProjectProperty, ProjectItem>(itemElement.Include, _data.Expander, itemElement.IncludeLocation);
+            var excludeItemspec = new ItemSpec<ProjectProperty, ProjectItem>(itemElement.Exclude, _data.Expander, itemElement.ExcludeLocation);
 
-                    // take out item references, we can't reason about them
-                    set.RemoveWhere(IsItemReferenceFragment);
+            var excludeSet = new Lazy<IEnumerable<string>>(() => excludeItemspec.Fragments.Where(f => !(f is ItemExpressionFragment<ProjectProperty, ProjectItem>)).Select(f => f.ItemSpecFragment).ToImmutableHashSet());
```

Comment:
Retrieve all the itemspec fragments of the exclude that are not item expressions.